### PR TITLE
Setting up Server & Database Connection

### PR DESCRIPTION
**Pull request by abbiehuynh**
## PR includes:
- **Code Structure:**
  - initial setup for the server 
  - organized code in server.js to separate route definitions from the database logic for better maintainability
  - included logging for debugging purposes to capture executed queries and their parameters
- **Database Connection Setup:**
  - configured the PostgreSQL connection using the pg library  
  - created a connection pool to manage database connections efficiently
  - added error handleing for connection failures
- **New Get Endpoint:**
  - added a new route /projectsByUserId/:user_id to fetch details about a specific user and all of their associated projects
  - added a new route /projectsByUser/:user_id/projects/:project_id to fetch all the details of a specific individual project of the user's
  - utilized LEFT JOIN to combine data from the users, projects, yarn, other materials, pattern, and images tables

### Testing: 
- tested the new endpoints using Postman to ensure it returns the correct data structure ofr valid user_id and project_id

## Testing Instructions:
To test the server, clone the repository 
```
https://github.com/abbiehuynh/my-crochet-corner.git <NameNewDirectory>
```
and run the following commands in the server:
1. `npm install`
2. `npm start`
3. Open http://localhost:3001 in your browser
- example of testing endpoint /projectsByUserId/:user_id :
  - http://localhost:3001/projectsByUserId/1
- example of testing endpoint /projectsByUser/:user_id/projects/:project_id
  - http://localhost:3001/projectsByUser/1/projects/2

### Notes:
Please review the code structure and suggest any improvements. 

---
Merging server branch into app-setup branch

### CODE REVIEW

File: server/db.sql [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+--
+
+ALTER TABLE ONLY public.users ALTER COLUMN id SET DEFAULT nextval('public.users_id_seq'::regclass);
+
+--
+-- Insert Values into users table
+INSERT INTO public.users(name, email, password, pronouns, bio, profile_picture) 
+ VALUES 
+    ('Abbie', 'abbiehuynhh@gmail.com', 'password', 'she/her', 'I love to crochet!', '' ),
+    ('Midnight', 'midnight@yahoo.com', 'coolbeans', 'he/him', 'String is cool.', '');
```
Comment (davidmedina-contact) [inline server/db.sql:67]: Understood this is temporary to set up the project. But adding a comment to remove this later is good practice, especially when there is sensitive data hardcoded

File: server/db.sql [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+INSERT INTO public.other_materials(project_id, hook_size, safety_eyes, stuffing) 
+ VALUES 
+    (2, '4mm', '10mm', '50 grams');
+
+--
+-- IMAGES -- MANY-TO-ONE -> PROJECTS
+-- Name: images; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE public.images (
```
Comment (davidmedina-contact) [inline server/db.sql:266]: Probably not for MPV: Is this table only for project images? Can this be more generic and be images for projects, yarns, patterns, etc?

File: server/package-lock.json
Comment (davidmedina-contact) [inline server/package-lock.json:1]: You don't need to commit this file

File: server/server.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+// middleware for parsing application/json
+app.use(cors());
+app.use(express.json());
+app.use(bodyParser.json());
+
+// test - creates an endpoint for the route  "/"
+app.get('/', (req, res) => {
+    res.json({ message: 'Hello from ExpressJS with React-Vite' });
+});
```
Comment (davidmedina-contact) [inline server/server.js:22]: Why is this needed?

File: server/server.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            LEFT JOIN
+                other_materials o ON p.id = o.project_id
+            LEFT JOIN 
+                pattern pa ON p.id = pa.project_id
+            LEFT JOIN
+                images i ON p.id = i.project_id
+            WHERE 
+                u.id = $1 AND p.id = $2;`, [userId, projectId]
+        );
+        res.send(projectsByUserId);
```
Comment (davidmedina-contact) [inline server/server.js:113]: How is this response structured? Is all the info returned on the same level or is there some grouping? This might be fine for the MVP but to make the app more modular and scalable, it might be better to structure the endpoints to return table-specific info; this way you'll have multiple endpoints with a single responsibility of fetching table data

**Review by davidmedina-contact — COMMENTED** [review]
Good progress! It might be worth thinking or documenting where are you making some design trade-offs for the MVP that you might want to improve on later versions of the app

File: server/db.sql (hunk 1 above)
Comment (abbiehuynh) [reply server/db.sql:67]: That is a great comment to add for my future self as a reminder to hide sensitive data, thank you.

File: server/db.sql (hunk 2 above)
Comment (abbiehuynh) [reply server/db.sql:266]: The images table is a many-to-one table with the projects table, but it is intended to be used for anything the user would like to include in their project photo album. However, to make the images table be more generic, would you suggest including the IDs of the other tables to be referenced in the images table?

File: server/server.js (hunk 3 above)
Comment (abbiehuynh) [reply server/server.js:22]: It was used to test the connection of the server to the client, by passing the message as a data prop to be render on the webpage. It will be removed after this PR has been documented.

File: server/db.sql (hunk 2 above)
Comment (davidmedina-contact) [reply server/db.sql:266]: Any table that uses an image can have an imageID array to reference one or more images. So the project, yarn and pattern tables can all have an imageID column/field to know what images they use

File: server/server.js (hunk 3 above)
Comment (davidmedina-contact) [reply server/server.js:22]: I see. Maybe worth adding a TODO comment here to remove later

File: server/server.js (hunk 4 above)
Comment (abbiehuynh) [reply server/server.js:113]: There is some grouping to the response depending on what and how much information is added to the tables. I intend to look into improving the structure of the endpoints after completing the mvp. I am interested in looking into how to restructure/ return table-specific info with multiple endpoints; thank you for the feedback and this insight.

File: server/db.sql (hunk 2 above)
Comment (abbiehuynh) [reply server/db.sql:266]: That makes more sense, thank you for the explanation! Since it is not for the mvp, I will make a note to look into adding the imageID array references when adding the "add image" feature. Thank you for the input.

**Comment by abbiehuynh** [comment]
@davidmedina-contact I've added the following changes based on the previous suggestions. Please let me know if there are still changes to be improved.

**Comment by davidmedina-contact** [comment]
LGTM
